hpx/parallel/util/numa_binding_allocator.hpp

```diff
@@ -0,0 +1,569 @@
+                         numa_binding_helper_ptr helper, size_type numa_domain,
+                         const std::vector<threads::hwloc_bitmap_ptr> &nodesets) const
+        {
+            const size_type pagesize  = sysconf(_SC_PAGE_SIZE);
+            const size_type pageN     = pagesize/sizeof(T);
+            const size_type num_pages = (n*sizeof(T) + pagesize - 1) / pagesize;
+            pointer page_ptr = p;
+            std::intptr_t memory = reinterpret_cast<std::intptr_t>(p);
+            HPX_ASSERT(memory % pagesize == 0);
+
+            LOG_NUMA_MSG("touch pages for numa " << numa_domain);
+            for (size_type i=0; i<num_pages; ++i) {
+                // we pass the base pointer and current page pointer
+                size_type dom = helper->operator()(p, page_ptr, pagesize, nodesets.size());
+                if (dom==numa_domain) {
+                    HPX_ASSERT( (std::size_t(page_ptr) & 4095) ==0 );
+                    // trigger a memory read and rewrite without changing contents
+                    volatile T* vaddr = const_cast<volatile T*>(page_ptr);
+                    *vaddr = *vaddr;
+#ifdef NUMA_BINDING_ALLOCATOR_INIT_MEMORY
+#if defined(NUMA_ALLOCATOR_LINUX)
+                    int Vmem = sched_getcpu(); // show which cpu is actually being used
+#else
+                    int Vmem = numa_domain;    // show just the domain we think we're on
+#endif
+                     pointer elem_ptr = page_ptr;
+                     for (size_type j=0; j<pageN; ++j) {
+                         *elem_ptr++ = T(Vmem);
+                     }
+ #endif
+                }
+                page_ptr += pageN;
+            }
+        }
+
+        // This is obsolete but kept for possible future use
+        void bind_pages(pointer p, size_t n,
+                        numa_binding_helper_ptr helper, size_type numa_domain,
+                        const std::vector<threads::hwloc_bitmap_ptr> &nodesets) const
+        {
+            const size_type pagesize  = sysconf(_SC_PAGE_SIZE);
+            const size_type pageN     = pagesize/sizeof(T);
+            const size_type num_pages = (n*sizeof(T) + pagesize - 1) / pagesize;
+            pointer page_ptr = p;
+            std::intptr_t memory = reinterpret_cast<std::intptr_t>(p);
+            HPX_ASSERT(memory % pagesize == 0);
+
+            LOG_NUMA_MSG("bind pages for numa " << numa_domain);
+            for (size_type i=0; i<num_pages; ++i) {
+                // we pass the base pointer and current page pointer
+                size_type dom = helper->operator()(p, page_ptr, pagesize, nodesets.size());
+                if (dom==numa_domain) {
+                    threads::topology().
+                        set_area_membind_nodeset(page_ptr, pagesize,
+                                                 nodesets[dom]->get_bmp());
+                }
+                page_ptr += pageN;
+            }
+        }
+
+    private:
+        hpx::lcos::local::spinlock display_mutex;
```

Comment:
This seems to be only needed if `NUMA_BINDING_ALLOCATOR_DEBUG_PAGE_BINDING` is defined